### PR TITLE
Support variable length match pattern from 1 to n steps

### PR DESCRIPTION
**Pull request by yixinglu**
depends on vesoft-inc/nebula-common#306

## Example

```text
> MATCH p= (:player{name:"Tim Duncan"})-[e:like*2..3{likeness:95}]-(v) return p
+-----------------------------------------------------------------------------------------------+
| p                                                                                             |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")<-[like@0]-("Tony Parker")-[like@0]->("Manu Ginobili")<-[like@0]-("Tim Duncan") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Tony Parker")-[like@0]->("Manu Ginobili")<-[like@0]-("Tim Duncan") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")<-[like@0]-("Tony Parker")<-[like@0]-("Tim Duncan")-[like@0]->("Manu Ginobili") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Tony Parker")-[like@0]->("Tim Duncan")-[like@0]->("Manu Ginobili") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Manu Ginobili")<-[like@0]-("Tony Parker")<-[like@0]-("Tim Duncan") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Manu Ginobili")<-[like@0]-("Tony Parker")-[like@0]->("Tim Duncan") |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")<-[like@0]-("Tony Parker")<-[like@0]-("Tim Duncan")                             |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")<-[like@0]-("Tony Parker")-[like@0]->("Manu Ginobili")                          |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Tony Parker")-[like@0]->("Manu Ginobili")                          |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Tony Parker")-[like@0]->("Tim Duncan")                             |
+-----------------------------------------------------------------------------------------------+
| ("Tim Duncan")-[like@0]->("Manu Ginobili")<-[like@0]-("Tony Parker")                          |
+-----------------------------------------------------------------------------------------------+
Got 11 rows (time spent 60353/62048 us)

Thu, 19 Nov 2020 17:28:34 CST
```

## Execution Plan

```text
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  id | name         | dependencies | profiling data                                        | operator info
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  36 | Project      | 35           | ver: 0, rows: 11, execTime: 550us, totalTime: 556us   | outputVar: [{"colNames":["p"],"name":"__Project_36","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_35
     |              |              |                                                       | columns: $p
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  35 | Filter       | 34           | ver: 0, rows: 11, execTime: 208us, totalTime: 214us   | outputVar: [{"colNames":["e","v","p"],"name":"__Filter_35","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_34
     |              |              |                                                       | condition: (cyclePath($-.p)==false)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  34 | Project      | 33           | ver: 0, rows: 11, execTime: 1277us, totalTime: 1283us | outputVar: [{"colNames":["e","v","p"],"name":"__Project_34","type":"DATASET"}]
     |              |              |                                                       | inputVar: __DataJoin_33
     |              |              |                                                       | columns: relationships($-._path_0) AS e,startNode($-._path_1) AS v,PathBuild[$-._path_0,$-._path_1] AS p
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  33 | DataJoin     | 32           | ver: 0, rows: 11, execTime: 1272us, totalTime: 1277us | outputVar: [{"colNames":["_path_0","_path_1"],"name":"__DataJoin_33","type":"DATASET"}]
     |              |              |                                                       | inputVar: {"rightVar":{"__Project_32":"0"},"leftVar":{"__Filter_28":"0"}}
     |              |              |                                                       | hashKeys: ["endNode($-._path)._vid"]
     |              |              |                                                       | probeKeys: ["startNode($-._path)._vid"]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  32 | Project      | 31           | ver: 0, rows: 3, execTime: 161us, totalTime: 167us    | outputVar: [{"colNames":["_path"],"name":"__Project_32","type":"DATASET"}]
     |              |              |                                                       | inputVar: __GetVertices_31
     |              |              |                                                       | columns: PathBuild[VERTEX]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  31 | GetVertices  | 30           | ver: 0, rows: 3, execTime: 267us, totalTime: 1686us   | outputVar: [{"colNames":[],"name":"__GetVertices_31","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Dedup_30
     |              |              |                                                       | space: 625
     |              |              |                                                       | dedup: false
     |              |              |                                                       | limit: 9223372036854775807
     |              |              |                                                       | filter:
     |              |              |                                                       | orderBy: []
     |              |              |                                                       | src: $-._vid
     |              |              |                                                       | props: []
     |              |              |                                                       | exprs: []
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  30 | Dedup        | 29           | ver: 0, rows: 3, execTime: 31us, totalTime: 35us      | outputVar: [{"colNames":["_vid"],"name":"__Dedup_30","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_29
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  29 | Project      | 28           | ver: 0, rows: 11, execTime: 215us, totalTime: 220us   | outputVar: [{"colNames":["_vid"],"name":"__Project_29","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_28
     |              |              |                                                       | columns: endNode($-._path)._vid
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  28 | Filter       | 27           | ver: 0, rows: 11, execTime: 247us, totalTime: 256us   | outputVar: [{"colNames":["_path"],"name":"__Filter_28","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Union_27
     |              |              |                                                       | condition: (length($-._path)>=2)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  27 | Union        | 26,17        | ver: 0, rows: 14, execTime: 20us, totalTime: 23us     | outputVar: [{"colNames":["_path"],"name":"__Union_27","type":"DATASET"}]
     |              |              |                                                       | leftVar: __Filter_25
     |              |              |                                                       | rightVar: __Union_17
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  26 | PassThrough  | 25           | ver: 0, rows: 0, execTime: 0us, totalTime: 2us        | outputVar: [{"colNames":["_path"],"name":"__Filter_25","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_25
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  25 | Filter       | 24           | ver: 0, rows: 6, execTime: 235us, totalTime: 240us    | outputVar: [{"colNames":["_path"],"name":"__Filter_25","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_24
     |              |              |                                                       | condition: (cyclePath($-._path)==false)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  24 | Project      | 23           | ver: 0, rows: 13, execTime: 854us, totalTime: 860us   | outputVar: [{"colNames":["_path"],"name":"__Project_24","type":"DATASET"}]
     |              |              |                                                       | inputVar: __DataJoin_23
     |              |              |                                                       | columns: PathBuild[$-._path_0,$-._path_1]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  23 | DataJoin     | 22           | ver: 0, rows: 13, execTime: 1302us, totalTime: 1308us | outputVar: [{"colNames":["_path_0","_path_1"],"name":"__DataJoin_23","type":"DATASET"}]
     |              |              |                                                       | inputVar: {"rightVar":{"__Project_22":"0"},"leftVar":{"__Filter_15":"0"}}
     |              |              |                                                       | hashKeys: ["endNode($-._path)._vid"]
     |              |              |                                                       | probeKeys: ["startNode($-._path)._vid"]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  22 | Project      | 21           | ver: 0, rows: 8, execTime: 402us, totalTime: 408us    | outputVar: [{"colNames":["_path"],"name":"__Project_22","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_21
     |              |              |                                                       | columns: PathBuild[VERTEX,EDGE] AS _path
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  21 | Filter       | 20           | ver: 0, rows: 8, execTime: 234us, totalTime: 240us    | outputVar: [{"colNames":[],"name":"__Filter_21","type":"DATASET"}]
     |              |              |                                                       | inputVar: __GetNeighbors_20
     |              |              |                                                       | condition: (EDGE.likeness==95)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  20 | GetNeighbors | 19           | ver: 0, rows: 25, execTime: 269us, totalTime: 4300us  | outputVar: [{"colNames":[],"name":"__GetNeighbors_20","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Dedup_19
     |              |              |                                                       | space: 625
     |              |              |                                                       | dedup: false
     |              |              |                                                       | limit: 9223372036854775807
     |              |              |                                                       | filter:
     |              |              |                                                       | orderBy: []
     |              |              |                                                       | src: $-._vid
     |              |              |                                                       | edgeTypes: []
     |              |              |                                                       | edgeDirection: BOTH
     |              |              |                                                       | vertexProps: []
     |              |              |                                                       | edgeProps: [{"type":"-629"},{"type":"629"}]
     |              |              |                                                       | statProps:
     |              |              |                                                       | exprs:
     |              |              |                                                       | random: false
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  19 | Dedup        | 18           | ver: 0, rows: 3, execTime: 19us, totalTime: 26us      | outputVar: [{"colNames":["_vid"],"name":"__Dedup_19","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_18
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  18 | Project      | 16           | ver: 0, rows: 5, execTime: 127us, totalTime: 133us    | outputVar: [{"colNames":["_vid"],"name":"__Project_18","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_15
     |              |              |                                                       | columns: endNode($-._path)._vid
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  16 | PassThrough  | 15           | ver: 0, rows: 0, execTime: 0us, totalTime: 2us        | outputVar: [{"colNames":["_path"],"name":"__Filter_15","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_15
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  15 | Filter       | 14           | ver: 0, rows: 5, execTime: 152us, totalTime: 158us    | outputVar: [{"colNames":["_path"],"name":"__Filter_15","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_14
     |              |              |                                                       | condition: (cyclePath($-._path)==false)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  14 | Project      | 13           | ver: 0, rows: 8, execTime: 505us, totalTime: 511us    | outputVar: [{"colNames":["_path"],"name":"__Project_14","type":"DATASET"}]
     |              |              |                                                       | inputVar: __DataJoin_13
     |              |              |                                                       | columns: PathBuild[$-._path_0,$-._path_1]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  13 | DataJoin     | 12           | ver: 0, rows: 8, execTime: 1024us, totalTime: 1031us  | outputVar: [{"colNames":["_path_0","_path_1"],"name":"__DataJoin_13","type":"DATASET"}]
     |              |              |                                                       | inputVar: {"rightVar":{"__Project_12":"0"},"leftVar":{"__Project_6":"0"}}
     |              |              |                                                       | hashKeys: ["endNode($-._path)._vid"]
     |              |              |                                                       | probeKeys: ["startNode($-._path)._vid"]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  12 | Project      | 11           | ver: 0, rows: 5, execTime: 272us, totalTime: 277us    | outputVar: [{"colNames":["_path"],"name":"__Project_12","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_11
     |              |              |                                                       | columns: PathBuild[VERTEX,EDGE] AS _path
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  11 | Filter       | 10           | ver: 0, rows: 5, execTime: 156us, totalTime: 162us    | outputVar: [{"colNames":[],"name":"__Filter_11","type":"DATASET"}]
     |              |              |                                                       | inputVar: __GetNeighbors_10
     |              |              |                                                       | condition: (EDGE.likeness==95)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  10 | GetNeighbors | 9            | ver: 0, rows: 13, execTime: 254us, totalTime: 3477us  | outputVar: [{"colNames":[],"name":"__GetNeighbors_10","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Dedup_9
     |              |              |                                                       | space: 625
     |              |              |                                                       | dedup: false
     |              |              |                                                       | limit: 9223372036854775807
     |              |              |                                                       | filter:
     |              |              |                                                       | orderBy: []
     |              |              |                                                       | src: $-._vid
     |              |              |                                                       | edgeTypes: []
     |              |              |                                                       | edgeDirection: BOTH
     |              |              |                                                       | vertexProps: []
     |              |              |                                                       | edgeProps: [{"type":"-629"},{"type":"629"}]
     |              |              |                                                       | statProps:
     |              |              |                                                       | exprs:
     |              |              |                                                       | random: false
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   9 | Dedup        | 8            | ver: 0, rows: 2, execTime: 15us, totalTime: 18us      | outputVar: [{"colNames":["_vid"],"name":"__Dedup_9","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_8
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   8 | Project      | 7            | ver: 0, rows: 3, execTime: 123us, totalTime: 128us    | outputVar: [{"colNames":["_vid"],"name":"__Project_8","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_6
     |              |              |                                                       | columns: endNode($-._path)._vid
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   7 | PassThrough  | 6            | ver: 0, rows: 0, execTime: 0us, totalTime: 2us        | outputVar: [{"colNames":["_path"],"name":"__Project_6","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_6
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   6 | Project      | 5            | ver: 0, rows: 3, execTime: 254us, totalTime: 260us    | outputVar: [{"colNames":["_path"],"name":"__Project_6","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_5
     |              |              |                                                       | columns: PathBuild[VERTEX,EDGE] AS _path
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   5 | Filter       | 4            | ver: 0, rows: 3, execTime: 154us, totalTime: 160us    | outputVar: [{"colNames":[],"name":"__Filter_5","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Filter_4
     |              |              |                                                       | condition: (EDGE.likeness==95)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   4 | Filter       | 3            | ver: 0, rows: 12, execTime: 166us, totalTime: 172us   | outputVar: [{"colNames":[],"name":"__Filter_4","type":"DATASET"}]
     |              |              |                                                       | inputVar: __GetNeighbors_3
     |              |              |                                                       | condition: (VERTEX.name==Tim Duncan)
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   3 | GetNeighbors | 2            | ver: 0, rows: 12, execTime: 300us, totalTime: 3481us  | outputVar: [{"colNames":[],"name":"__GetNeighbors_3","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Dedup_2
     |              |              |                                                       | space: 625
     |              |              |                                                       | dedup: false
     |              |              |                                                       | limit: 9223372036854775807
     |              |              |                                                       | filter:
     |              |              |                                                       | orderBy: []
     |              |              |                                                       | src: $-._vid
     |              |              |                                                       | edgeTypes: []
     |              |              |                                                       | edgeDirection: BOTH
     |              |              |                                                       | vertexProps: []
     |              |              |                                                       | edgeProps: [{"type":"-629"},{"type":"629"}]
     |              |              |                                                       | statProps:
     |              |              |                                                       | exprs:
     |              |              |                                                       | random: false
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   2 | Dedup        | 1            | ver: 0, rows: 1, execTime: 18us, totalTime: 22us      | outputVar: [{"colNames":["_vid"],"name":"__Dedup_2","type":"DATASET"}]
     |              |              |                                                       | inputVar: __Project_1
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
   1 | Project      | 39           | ver: 0, rows: 1, execTime: 111us, totalTime: 117us    | outputVar: [{"colNames":["_vid"],"name":"__Project_1","type":"DATASET"}]
     |              |              |                                                       | inputVar: __IndexScan_0
     |              |              |                                                       | columns: $_vid
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  39 | IndexScan    | 38           | ver: 0, rows: 1, execTime: 0us, totalTime: 1661us     | outputVar: [{"colNames":[],"name":"__IndexScan_0","type":"DATASET"}]
     |              |              |                                                       | inputVar:
     |              |              |                                                       | space: 625
     |              |              |                                                       | dedup: false
     |              |              |                                                       | limit: 9223372036854775807
     |              |              |                                                       | filter:
     |              |              |                                                       | orderBy: []
     |              |              |                                                       | schemaId: 626
     |              |              |                                                       | isEdge: false
     |              |              |                                                       | returnCols: []
     |              |              |                                                       | indexCtx: [{"columnHints":[{"endValue":"__EMPTY__","beginValue":"Tim Duncan","column":"name","scanType":"PREFIX"}],"index_id":632,"filter":""}]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  38 | Start        |              | ver: 0, rows: 0, execTime: 3us, totalTime: 48us       | outputVar: [{"colNames":[],"name":"__Start_38","type":"DATASET"}]
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
  17 | Union        | 16,7         | ver: 0, rows: 8, execTime: 23us, totalTime: 26us      | outputVar: [{"colNames":["_path"],"name":"__Union_17","type":"DATASET"}]
     |              |              |                                                       | leftVar: __Filter_15
     |              |              |                                                       | rightVar: __Project_6
-----+--------------+--------------+-------------------------------------------------------+--------------------------------------------------------------------------------------------------------------------------------------------------
```

plan

### CODE REVIEW

File: src/planner/planners/MatchSolver.h [hunk 1]
```diff
@@ -22,6 +25,21 @@ class MatchSolver final {
...
+    static Expression* makeIndexFilter(const std::string& label,
+                                       const MapExpression* map,
+                                       QueryContext* qctx);
+
+    static Expression* makeIndexFilter(const std::string& label,
+                                       const std::string& alias,
+                                       Expression* filter,
+                                       QueryContext* qctx);
+
+    static Status buildFilter(const MatchAstContext* mctx, SubPlan* plan);
```
Comment (Shylock-Hg) [inline src/planner/planners/MatchSolver.h:42]: Don't mix use `build` and `make`.

Comment (yixinglu) [reply src/planner/planners/MatchSolver.h:42]: Only move existing codes here, well, I will fix it in next zero to one step PR.

**Review by CPWstatic — APPROVED** [review]

File: src/planner/planners/MatchVariableLengthPatternIndexScanPlanner.cpp [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+    auto columns = saveObject(new YieldColumns);
+    auto pathExpr = std::make_unique<PathBuildExpression>();
+    pathExpr->add(std::make_unique<VertexExpression>());
+    columns->addColumn(new YieldColumn(pathExpr.release()));
+    plan->root = Project::make(qctx, root, columns);
+    plan->root->setColNames({kPath});
+    return Status::OK();
+}
+
+Status MatchVariableLengthPatternIndexScanPlanner::filterDatasetByPathLength(const NodeInfo &node,
```
Comment (Shylock-Hg) [inline src/planner/planners/MatchVariableLengthPatternIndexScanPlanner.cpp:279]: Don't need to filter if don't union the data less than min hop.

Comment (yixinglu) [reply src/planner/planners/MatchVariableLengthPatternIndexScanPlanner.cpp:279]: Right, But I don't want to optimize this case here. If so, our plan generated code will be filled by many branches. I prefer to add some optimization rules to handle it.

Comment (Shylock-Hg) [reply src/planner/planners/MatchVariableLengthPatternIndexScanPlanner.cpp:279]: Make sense, but this pattern seems not easy to match in optimizer.

**Review by nevermore3 — APPROVED** [review]

**Pull request merged**